File: packages/tanstack-workflow-runtime/src/define-runtime.ts
// @ts-nocheck -- Vendored TanStack source targets a looser optional-property TypeScript policy.
import { createRuntimeDriver } from "./runtime-driver"
import type {
  WorkflowRegistrationMap,
  WorkflowRuntimeConfig,
  WorkflowRuntimeDefinition,
  WorkflowScheduleSpec
} from "./types"

export function defineWorkflowRuntime<
  const TWorkflows extends WorkflowRegistrationMap
>(
  config: WorkflowRuntimeConfig<TWorkflows>
): WorkflowRuntimeDefinition<TWorkflows> {
  const driver = createRuntimeDriver(config)
  return {
    __kind: "workflow-runtime",
    ...config,
    ...driver
  }
}

export function cron(
  expression: string,
  options: { timezone?: string } = {}
): WorkflowScheduleSpec {
  return {
    kind: "cron",
    expression,
    timezone: options.timezone
  }
}

export const every = {
  milliseconds(everyMs: number): WorkflowScheduleSpec {
    return { kind: "interval", everyMs }
  },
  seconds(seconds: number): WorkflowScheduleSpec {
    return { kind: "interval", everyMs: seconds * 1000 }
  },
  minutes(minutes: number): WorkflowScheduleSpec {
    return { kind: "interval", everyMs: minutes * 60 * 1000 }
  },
  hours(hours: number): WorkflowScheduleSpec {
    return { kind: "interval", everyMs: hours * 60 * 60 * 1000 }
  }
}

File: src/Firegrid.Runtime/DefineRuntime.fs
namespace Firegrid.Runtime

open Firegrid.Core

/// `define-runtime.ts` — `defineWorkflowRuntime`, `cron`, `every`.
[<RequireQualifiedAccess>]
module DefineRuntime =

    /// `defineWorkflowRuntime(config)`.
    let defineWorkflowRuntime (config: WorkflowRuntimeConfig) : WorkflowRuntimeDefinition =
        let driver = RuntimeDriver.create config

        { Kind = "workflow-runtime"
          Workflows = config.Workflows
          Store = config.Store
          DefaultLeaseMs = config.DefaultLeaseMs
          StartRun = driver.StartRun
          DeliverSignal = driver.DeliverSignal
          DeliverApproval = driver.DeliverApproval
          Sweep = driver.Sweep }

    /// `cron(expression, { timezone? })`.
    let cron (expression: string) (timezone: string option) : WorkflowScheduleSpec = CronSpec(expression, timezone)

    /// `every` — interval helpers.
    [<RequireQualifiedAccess>]
    module every =

        let milliseconds (everyMs: float) : WorkflowScheduleSpec = IntervalSpec(everyMs, None)

        let seconds (seconds: float) : WorkflowScheduleSpec = IntervalSpec(seconds * 1000.0, None)

        let minutes (minutes: float) : WorkflowScheduleSpec = IntervalSpec(minutes * 60.0 * 1000.0, None)

        let hours (hours: float) : WorkflowScheduleSpec = IntervalSpec(hours * 60.0 * 60.0 * 1000.0, None)
